http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -416,4 +484,112 @@ private static boolean hasScope(String scopeParam, String targetScope) {
+        forwardedQueryParams.add(new BasicNameValuePair(CLIENT_ID, deployment.getResourceName()));
+
+        JwtClaims jwtClaims = new JwtClaims();
+        jwtClaims.setIssuer(deployment.getIssuerUrl());
+        jwtClaims.setAudience(deployment.getProviderUrl());
+        for ( NameValuePair parameter: forwardedQueryParams) {
+            jwtClaims.setClaim(parameter.getName(), parameter.getValue());
+        }
+
+        // sign JWT first before encrypting
+        JsonWebSignature signedRequest = signRequest(jwtClaims);
+
+        // Encrypting optional
+        if (deployment.getRequestEncryptAlgorithm() != null && !deployment.getRequestEncryptAlgorithm().isEmpty() &&
+            deployment.getRequestEncryptEncValue() != null && !deployment.getRequestEncryptEncValue().isEmpty()) {
+            return encryptRequest(signedRequest).getCompactSerialization();
+        } else {
+            return signedRequest.getCompactSerialization();
+        }
+    }
+
+    public KeyPair getkeyPair() throws IOException {
+        if (deployment.getClientKeyStoreFile().contains(PROTOCOL_CLASSPATH)) {
+            deployment.setClientKeyStoreFile(deployment.getClientKeyStoreFile().replace(PROTOCOL_CLASSPATH, Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("")).getPath()));
+        }
+        if (!deployment.getRequestSignatureAlgorithm().equals(NONE) && deployment.getClientKeyStoreFile() == null){
+            throw log.invalidKeyStoreConfiguration();
+        } else {
+            return loadKeyPairFromKeyStore(deployment.getClientKeyStoreFile(),
+                    deployment.getClientKeyStorePassword(), deployment.getClientKeyPassword(),
+                    deployment.getClientKeyAlias(), deployment.getClientKeystoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws Exception {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (deployment.getRequestSignatureAlgorithm().contains(NONE)) { //unsigned
+            jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
+            jsonWebSignature.setAlgorithmHeaderValue(NONE);
+        } else if (deployment.getRequestSignatureAlgorithm().contains(HS_256)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_384)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_512)) { //signed with symmetric key
+            jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestSignatureAlgorithm());
+            Key key = new HmacKey(deployment.getResourceCredentials().get("secret").toString().getBytes(StandardCharsets.UTF_8));   //the client secret is a shared secret between the server and the client
+            jsonWebSignature.setDoKeyValidation(false);     //skips validation so that size of the secret does not matter
```

Comment:
Why is this needed?

Reply:
This is because OpenID providers, keycloak included can support Symmetric keys to sign the JWT. In these cases, I have been able to sign it using the client secret since that is a secret key shared between both the client and the OpenID provider. 

I tried to generate another secret key and upload it to keycloak, but it did not work since keycloak stores the certificate containing a public key.

Reply:
Oh and if you are asking about why we are skipping validation, that is because otherwise, the size of the secret key must match the size of the algorithm. And since I can't change the size of the secret key, we are skipping the size validation.